lib/renderer/src/renderer/ResourceStorage.h

```diff
@@ -0,0 +1,86 @@
+//
+// Created by nikita on 26.11.2021.
+//
+
+#ifndef VANGERS_RIDSTORAGE_H
+#define VANGERS_RIDSTORAGE_H
+
+#include <unordered_map>
+#include <memory>
+
+#include "common.h"
+#include "exception.h"
+
+namespace renderer {
+	template<typename TResourceId, typename TResource>
+	class ResourceStorage {
+	private:
+		static const std::hash<int32_t> hash;
+	public:
+		struct ResourceIdHash {
+			size_t operator()(TResourceId rid) const noexcept{
+				return hash(rid.id);
+			}
+		};
+
+		struct ResourceIdEquals {
+			bool operator()(const TResourceId& rid1, const TResourceId& rid2) const{
+				return rid1.id == rid2.id;
+			}
+		};
+
+
+		typedef typename std::unordered_map<TResourceId, std::unique_ptr<TResource>, ResourceIdHash, ResourceIdEquals>::iterator iterator;
+
+		ResourceStorage()
+			:_next(0)

```

Comment:
I think we should never use the internal `id` field explicitly, so zero should be fine. Anyway I plan to refactor the "resource" system, since it's a bit over engineered..